jetty-servlet/src/test/java/org/eclipse/jetty/servlet/GzipHandlerCommitTest.java

```diff
@@ -0,0 +1,157 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.servlet;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Arrays;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.servlet.ServletException;
+import javax.servlet.ServletOutputStream;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.ServerConnector;
+import org.eclipse.jetty.server.handler.gzip.GzipHandler;
+import org.eclipse.jetty.util.component.LifeCycle;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.lessThan;
+
+public class GzipHandlerCommitTest
+{
+    private static Server server;
+    private static HttpClient client;
+
+    @BeforeEach
+    public void startUp() throws Exception
+    {
+        server = new Server();
+        ServerConnector connector = new ServerConnector(server);
+        connector.setPort(0);
+        server.addConnector(connector);
+
+        ServletContextHandler contextHandler = new ServletContextHandler();
+        contextHandler.setContextPath("/");
+        contextHandler.addServlet(FlushBufferServlet.class, "/flush-buffer/*");
+
+        GzipHandler gzipHandler = new GzipHandler();
+        gzipHandler.setHandler(contextHandler);
+
+        server.setHandler(gzipHandler);
+        server.start();
+
+        client = new HttpClient();
+        client.start();
+    }
+
+    @AfterEach
+    public void tearDown()
+    {
+        LifeCycle.stop(client);
+        LifeCycle.stop(server);
+    }
+
+    /**
+     * A servlet should be able to flush and then produce no content.
+     */
+    @Test
+    public void testFlushNoContent() throws Exception
+    {
+        long delay = 2000;
```

Comment:
These tests could be rewritten easily so that each test passes its own `Servlet` implementation to the `Server` startup.
Then, it would be easy to use a latch to coordinate with the client, so that the test does not need to wait 2000 ms but it will be immediate.

Pseudo code:

```java
@Test
public void test() {
  CountDownLatch latch = new CountDownLatch(1);
  start(new HttpServlet() {
    service(request, response) {
      response.flushBuffer();
      latch.await();
    }
  });

  client.newRequest(...)
    .onResponseHeaders(response -> latch.countDown())
    ...
    .send();
}
```

Reply:
This is an invalid test unfortunately.
If we just use latch.await() then we'd have to rely on client.setIdleTimeout().
If we do that, the client finishes and errors out with TimeoutException, and the servlet is now hung and the server.stop() doesn't clean it up in a timely fashion.
Also, if we switch to a latch.await(timeout), the flush will occur when the servlet is done with its dispatch, which is a false positive.


We want to hold the dispatch and test the timing.
As the bug is that the flush doesn't immediately flush when requested.
The bug isn't the flush never occurs. (which your proposed test is good for)